src/sentry_session.c

```diff
@@ -0,0 +1,225 @@
+    if (sentry_value_is_null(value)) {
+        return NULL;
+    }
+
+    sentry_session_t *rv = SENTRY_MAKE(sentry_session_t);
+    rv->session_id
+        = sentry__value_as_uuid(sentry_value_get_by_key(value, "sid"));
+
+    sentry_value_t distinct_id = sentry_value_get_by_key(value, "did");
+    if (sentry_value_is_null(distinct_id)) {
+        rv->distinct_id = NULL;
+    } else {
+        rv->distinct_id
+            = sentry__string_clone(sentry_value_as_string(distinct_id));
+    }
+
+    const char *status
+        = sentry_value_as_string(sentry_value_get_by_key(value, "status"));
+    rv->status = status_from_string(status);
+
+    rv->init = sentry_value_is_true(sentry_value_get_by_key(value, "init"));
+    rv->errors = (int64_t)sentry_value_as_int32(
+        sentry_value_get_by_key(value, "errors"));
+    rv->duration_ms = (uint64_t)(
+        sentry_value_as_double(sentry_value_get_by_key(value, "duration"))
+        * 1000);
+    rv->started_ms = (uint64_t)(
+        sentry_value_as_double(sentry_value_get_by_key(value, "started"))
+        * 1000);
+
+    return rv;
+}
+
+sentry_session_t *
+sentry__session_from_path(const sentry_path_t *path)
+{
+    size_t buf_len;
+    char *buf = sentry__path_read_to_buffer(path, &buf_len);
+    if (!buf) {
+        return NULL;
+    }
+
+    sentry_session_t *rv = sentry__session_from_json(buf, buf_len);
+    sentry_free(buf);
+    return rv;
+}
+
+void
+sentry_start_session(void)
+{
+    sentry_end_session();
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        scope->session = sentry__session_new();
+        sentry__scope_session_sync(scope);
+    }
+}
+
+void
+sentry__end_current_session_with_status(sentry_session_status_t status)
+{
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        if (scope->session) {
+            scope->session->status = status;
+        }
+    }
+    sentry_end_session();
+}
+
+void
+sentry__record_errors_on_current_session(uint32_t error_count)
+{
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        if (scope->session) {
+            scope->session->errors += error_count;
+        }
+    }
+}
+
+void
+sentry_end_session(void)
+{
+    sentry_envelope_t *envelope = NULL;
+
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        if (scope->session) {
+            if (scope->session->status == SENTRY_SESSION_STATUS_OK) {
+                scope->session->status = SENTRY_SESSION_STATUS_EXITED;
+            }
+            envelope = sentry__envelope_new();
+            sentry__envelope_add_session(envelope, scope->session);
+            sentry__session_free(scope->session);
+            scope->session = NULL;
+        }
+    }
+
+    if (envelope) {
+        sentry__capture_envelope(envelope);
+    }
+}
+
+void
+sentry__add_current_session_to_envelope(sentry_envelope_t *envelope)
+{
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        if (scope->session) {
+            sentry__envelope_add_session(envelope, scope->session);
+            scope->session->init = false;
```

Comment:
this is the only place you set this to `false`. I don’t really understand what this flag is for. can you add an explanation somewhere?

Reply:
The initial http request needs the flag to be set to `true`. If the session is reused afterwards it must be `false`.